staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -17,330 +17,239 @@ limitations under the License.
+	service := v1.APIService{}
+	service.Name = name
+	info := &openAPISpecInfo{
+		apiService: service,
 	}
-	sortByPriority(specs)
-	for _, specInfo := range specs {
-		if specToReturn == nil {
-			specToReturn = &spec.Swagger{}
-			*specToReturn = *specInfo.spec
-			// Paths and Definitions are set by MergeSpecsIgnorePathConflict
-			specToReturn.Paths = nil
-			specToReturn.Definitions = nil
-		}
-		if err := aggregator.MergeSpecsIgnorePathConflict(specToReturn, specInfo.spec); err != nil {
-			return nil, err
-		}
-	}
-
-	return specToReturn, nil
+	info.spec.Replace(spec)
+	s.specsByAPIServiceName[name] = info
 }
 
-// updateOpenAPISpec aggregates all OpenAPI specs.  It is not thread-safe. The caller is responsible to hold proper locks.
-func (s *specAggregator) updateOpenAPISpec() error {
-	if s.openAPIVersionedService == nil {
-		return nil
-	}
-	specToServe, err := s.buildOpenAPISpec()
-	if err != nil {
-		return err
-	}
-	return s.openAPIVersionedService.UpdateSpec(specToServe)
-}
-
-// tryUpdatingServiceSpecs tries updating openAPISpecs map with specified specInfo, and keeps the map intact
-// if the update fails.
-func (s *specAggregator) tryUpdatingServiceSpecs(specInfo *openAPISpecInfo) error {
-	if specInfo == nil {
-		return fmt.Errorf("invalid input: specInfo must be non-nil")
-	}
-	_, updated := s.openAPISpecs[specInfo.apiService.Name]
-	origSpecInfo, existedBefore := s.openAPISpecs[specInfo.apiService.Name]
-	s.openAPISpecs[specInfo.apiService.Name] = specInfo
-
-	// Skip aggregation if OpenAPI spec didn't change
-	if existedBefore && origSpecInfo != nil && origSpecInfo.etag == specInfo.etag {
-		return nil
-	}
-	klog.V(2).Infof("Updating OpenAPI spec because %s is updated", specInfo.apiService.Name)
-	defer func(start time.Time) {
-		duration := time.Since(start)
-		klog.V(2).Infof("Finished OpenAPI spec generation after %v", duration)
-
-		reason := "add"
-		if updated {
-			reason = "update"
-		}
-
-		regenerationCounter.With(map[string]string{"apiservice": specInfo.apiService.Name, "reason": reason})
-		regenerationDurationGauge.With(map[string]string{"reason": reason}).Set(duration.Seconds())
-	}(time.Now())
-	if err := s.updateOpenAPISpec(); err != nil {
-		if existedBefore {
-			s.openAPISpecs[specInfo.apiService.Name] = origSpecInfo
-		} else {
-			delete(s.openAPISpecs, specInfo.apiService.Name)
+// buildMergeSpecLocked creates a new cached mergeSpec from the list of cached specs.
+func (s *specAggregator) buildMergeSpecLocked() cached.Data[*spec.Swagger] {
+	apiServices := make([]*v1.APIService, 0, len(s.specsByAPIServiceName))
+	for k := range s.specsByAPIServiceName {
+		apiServices = append(apiServices, &s.specsByAPIServiceName[k].apiService)
+	}
+	sortByPriority(apiServices)
+	caches := make([]cached.Data[*spec.Swagger], len(s.specsByAPIServiceName))
+	for i, apiService := range apiServices {
+		caches[i] = &(s.specsByAPIServiceName[apiService.Name].spec)
```

Comment:
Nit, but it's interesting you make two lists in a row and you use two different patterns of `make` and iterations. Also, I know this is irrelevant, but you probably should use the length of the structure that you use for the range for clarity (i.e. line 153 you use `len(s.specByAPIServiceName)` and line 155 you range over apiServices, I know they are the same time, but it'd be clearer if they matched).

Reply:
Updated the `make` function to use the length of the structure. List 1 is technically constructed from a map and list 2 is constructed from a list, hence the slight mismatch in how they're populated.